heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/TableDryRunRender.java

```diff
@@ -0,0 +1,157 @@
+// Copyright 2016 Twitter. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.twitter.heron.scheduler.utils;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+import com.google.common.base.Strings;
+
+import com.twitter.heron.spi.packing.Resource;
+
+/**
+ * Table dry-run render
+ */
+public class TableDryRunRender extends DryRunRender {
+
+  private List<String> title = Arrays.asList(
+      "component", "cpu", "disk (GB)", "ram (GB)", "parallelism");
+
+  private StringBuilder addRow(StringBuilder builder, String row) {
+    builder.append(row);
+    builder.append('\n');
+    return builder;
+  }
+
+  /**
+   * generate formatter for each row based on rows. Width of a column is the
+   * max width of all cells on that column
+   *
+   * @param rows Each row in table
+   * @return formatter for row
+   */
+  private String generateRowFormatter(List<List<String>> rows) {
+    Integer[] width = new Integer[title.size()];
+    for (int i = 0; i < title.size(); i++) {
+      width[i] = title.get(i).length();
+    }
+    for (List<String> row: rows) {
+      for (int i = 0; i < row.size(); i++) {
+        width[i] = Math.max(width[i], row.get(i).length());
+      }
+    }
+    StringBuilder metaFormatterBuilder = new StringBuilder();
+    String metaCellFormatter = "%%%ds";
+    metaFormatterBuilder.append(Strings.repeat(String.format("| %s ", metaCellFormatter),
+        title.size()));
+    metaFormatterBuilder.append("|");
+    return String.format(metaFormatterBuilder.toString(), (Object[]) width);
+  }
+
+  /**
+   * Seal rows to create table
+   * @param rows Each row in table
+   * @return Formatted table
+   */
+  private String createTable(List<List<String>> rows) {
```

Comment:
I was wondering if we could use List and Joiner here instead of StringBuilder and addRow method as following code.

```
List<String> contentList = new ArrayList();
contentList.add(Strings.repeat("=", titleRow.length()));
contentList.add(titleRow);
contentList.add(Strings.repeat("-", titleRow.length()));
for (List<String> row: rows) {
     contentList.add(String.format(rowFormatter, (Object[]) row.toArray(new String[row.size()])));
}
contentList.add(Strings.repeat("=", titleRow.length()));
return Joiner.on("\n").join(contentList)
```

I'm not sure which one is better.